nifi-framework-bundle/nifi-framework-extensions/nifi-py4j-framework-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/componentstate.py

```diff
@@ -0,0 +1,171 @@
+        """
+        :param state: (dict) Python dictionary containing the state to be saved.
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        """
+        try:
+            self._java_state_manager.setState(convert_pydict_to_jmap(state), convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+    def getState(self, scope):
+        """
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        :return: Python StateMap object representing the state.
+        """
+        try:
+            java_statemap = self._java_state_manager.getState(convert_pyscope_to_jscope(scope))
+            return StateMap(java_statemap)
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+    def replace(self, oldValue, newValue, scope):
+        """
+        :param oldValue: (StateMap) Python StateMap object with the state to be replaced.
+        :param newValue: (dict) Python dictionary containing the state to be saved.
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        """
+        try:
+            self._java_state_manager.replace(oldValue.get_java_statemap(), convert_pydict_to_jmap(newValue), convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+    def clear(self, scope):
+        """
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        """
+        try:
+            self._java_state_manager.clear(convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
+
+
+class StateMap:
+    """
+    Python class wrapping the StateMap Java implementation.
+
+    The constructor receives a Java object implementing the Java StateMap interface.
+    The methods in Java StateMap have corresponding methods in this class.
+    Methods in this class propagate the method calls to the Java StateMap object.
+    """
+
+    _java_statemap = None
+
+    def __init__(self, java_statemap):
+        """
+        :param java_statemap: Java StateMap object.
+        """
+        self._java_statemap = java_statemap
+
+    def getStateVersion(self):
+        """
+        :return: (int) Integer value of the state version, or None if it is not set.
+        """
+        if self._java_statemap.getStateVersion().isEmpty():
+            return None
+        else:
+            return self._java_statemap.getStateVersion().get()
+
+    def get(self, key):
+        """
+        :param key: (str) String key indicating which state value should be returned.
+        :return: (str) String state value.
+        """
+        return self._java_statemap.get(key)
+
+    def toMap(self):
+        """
+        :return: (dict) Python dictionary representation of the state.
+        """
+        return self.to_dictionary()
+
+    def to_dictionary(self):
```

Comment:
Is there a reason for defining this method separately from `toMap`?

Reply:
I wanted to indicate that we are aware that it is not actually a Map, but a Dictionary, but the method name needs to (more precisely we want it to) match that of the Java class. I might have overthought it. Removed to_dictionary().